pkg/pipeline/encode/encode_prom.go

```diff
@@ -261,6 +263,217 @@ func (e *EncodeProm) cleanupExpiredEntriesLoop() {
 	}
 }
 
+func (e *EncodeProm) addCounter(fullMetricName string, mInfo *MetricInfo) {
+	counter := prometheus.NewCounterVec(prometheus.CounterOpts{Name: fullMetricName, Help: ""}, mInfo.Labels)
+	err := e.registerer.Register(counter)
+	if err != nil {
+		log.Errorf("error during prometheus.Register: %v", err)
+	}
+	e.counters[fullMetricName] = counterInfo{
+		counter: counter,
+		info:    mInfo,
+	}
+}
+
+func (e *EncodeProm) addGauge(fullMetricName string, mInfo *MetricInfo) {
+	gauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{Name: fullMetricName, Help: ""}, mInfo.Labels)
+	err := e.registerer.Register(gauge)
+	if err != nil {
+		log.Errorf("error during prometheus.Register: %v", err)
+	}
+	e.gauges[fullMetricName] = gaugeInfo{
+		gauge: gauge,
+		info:  mInfo,
+	}
+}
+func (e *EncodeProm) addHistogram(fullMetricName string, mInfo *MetricInfo) {
+	histogram := prometheus.NewHistogramVec(prometheus.HistogramOpts{Name: fullMetricName, Help: ""}, mInfo.Labels)
+	err := e.registerer.Register(histogram)
+	if err != nil {
+		log.Errorf("error during prometheus.Register: %v", err)
+	}
+	e.histos[fullMetricName] = histoInfo{
+		histo: histogram,
+		info:  mInfo,
+	}
+}
+func (e *EncodeProm) addAgghistogram(fullMetricName string, mInfo *MetricInfo) {
+	agghistogram := prometheus.NewHistogramVec(prometheus.HistogramOpts{Name: fullMetricName, Help: ""}, mInfo.Labels)
+	err := e.registerer.Register(agghistogram)
+	if err != nil {
+		log.Errorf("error during prometheus.Register: %v", err)
+	}
+	e.aggHistos[fullMetricName] = histoInfo{
+		histo: agghistogram,
+		info:  mInfo,
+	}
+}
+
+func (e *EncodeProm) cleanDeletedCounters(newCfg api.PromEncode) {
```

Comment:
after rebasing this code can probably be simplified, with all cleanup functions merged in a single one, as Kalman did use a common struct for all the metric types